File: ARRAYS/912. Sort an Array.java
//problem link: https://leetcode.com/problems/sort-an-array/
//timeComplexity: o(nlogn)
//spaceComplexity: o(n)

class Solution {
    public void merge(int low,int mid,int high,int nums[]){
        int i=low,j=mid+1;
        ArrayList<Integer>temp=new ArrayList<>();
        while(i<=mid && j<=high){
            if(nums[i]<=nums[j]){
                temp.add(nums[i]);
                i++;
            }
            else{
                temp.add(nums[j]);
                j++;
            }
        }
        while(i<=mid){
            temp.add(nums[i]);
            i++;
        }
        while(j<=high){
            temp.add(nums[j]);
            j++;
        }
        for(i=0;i<temp.size();i++){
            nums[i+low]=temp.get(i);
        }
    }
    public void mergeSort(int low,int high,int []nums){
        if(low<high){
            int mid=low+(high-low)/2;
            mergeSort(low,mid,nums);
            mergeSort(mid+1,high,nums);
            merge(low,mid,high,nums);
        }
    }
    public int[] sortArray(int[] nums) {
        int n=nums.length;
        mergeSort(0,n-1,nums);
        return nums;
    }
}

File: RECURSION/912. Sort an Array.cpp
//problem link: https://leetcode.com/problems/sort-an-array/?envType=problem-list-v2&envId=merge-sort
//timeComplexity: O(nlog(n))
//spaceComplexity: O(n)

class Solution {
    void merge(int low,int mid,int high,vector<int>&nums){
        int left=low,right=mid+1;
        vector<int>temp;
        while(left<=mid && right<=high){
            if(nums[left]<=nums[right]){
                temp.push_back(nums[left]);
                left++;
            }
            else{
                temp.push_back(nums[right]);
                right++;
            }
        }
        while(left<=mid){
            temp.push_back(nums[left]);
            left++;
        }
        while(right<=high){
            temp.push_back(nums[right]);
            right++;
        }
        for(int i=low;i<=high;i++){
            nums[i]=temp[i-low];
        }
        return;
    }
    void mergeSort(int low,int high,vector<int>&nums){
        if(low>=high)return;
        int mid=low+(high-low)/2;
        mergeSort(low,mid,nums);
        mergeSort(mid+1,high,nums);
        merge(low,mid,high,nums);
    }
public:
    vector<int> sortArray(vector<int>& nums) {
        int n=nums.size();
        mergeSort(0,n-1,nums);
        return nums;
    }
};